@xen-orchestra/fs/src/_encryptor.js

```diff
@@ -0,0 +1,136 @@
+const crypto = require('crypto')
+const { Transform } = require('stream')
+
+class Crypt extends Transform {
+  #initDone = false
+  #iv
+  #cipher
+  constructor(iv, cipher) {
+    super()
+    this.#iv = iv
+    this.#cipher = cipher
+  }
+
+  _transform(chunk, encoding, callback) {
+    if (!this.#initDone) {
+      this.push(this.#iv)
+      this.#initDone = true
+    }
+    const cryptedChunk = this.#cipher.update(chunk)
+    this.push(cryptedChunk)
+    callback()
+  }
+  _flush(callback) {
+    this.push(this.#cipher.final())
+    callback()
+  }
+
+  // handle a circular reference error when communicating with backup worker
+  toJSON() {
+    return {
+      iv: this.#iv,
+      cipher: this.cipher,
+    }
+  }
+}
+
+class Decrypt extends Transform {
+  #algorithm
+  #initDone = false
+  #ivLength
+  #key
+  #cipher
+  constructor({ algorithm, key, ivLength }) {
+    super()
+    this.#ivLength = ivLength
+    this.#algorithm = algorithm
+    this.#key = key
+  }
+
+  _transform(chunk, encoding, callback) {
+    if (!this.#initDone) {
+      const iv = chunk.slice(0, this.#ivLength)
+      this.#cipher = crypto.createDecipheriv(this.#algorithm, Buffer.from(this.#key), iv)
+      chunk = chunk.slice(this.#ivLength)
+      this.#initDone = true
+    }
+    const decryptedChunk = this.#cipher.update(chunk)
+    this.push(decryptedChunk)
+    callback()
+  }
+  _flush(callback) {
+    this.push(this.#cipher.final())
+    callback()
+  }
+
+  // handle a ciruclar reference error when communicating with backup worker
+  toJSON() {
+    return {
+      algorithm: this.#algorithm,
+    }
+  }
+}
+function getEncryptor(key = undefined) {
+  if (key === undefined) {
+    return {
+      id: 'NULL_ENCRYPTOR',
+      algorithm: 'none',
+      key: 'none',
+      ivLength: 0,
+      encryptData: buffer => buffer,
+      encryptStream: stream => stream,
+      decryptData: buffer => buffer,
+      decryptStream: stream => stream,
+    }
+  }
+  const algorithm = 'aes-256-cbc'
+  const ivLength = 16
+
+  function encryptStream(input) {
+    const iv = crypto.randomBytes(ivLength)
+    const cipher = crypto.createCipheriv(algorithm, Buffer.from(key), iv)
+
+    const cryptoStream = new Crypt(iv, cipher)
+
+    return input.pipe(cryptoStream)
```

Comment:
Avoid using `stream.pipe()`, use [`pumpify`](https://github.com/mafintosh/pumpify) instead :slightly_smiling_face:

Reply:
what are the benefits of using pumpify ?

Reply:
Streams are unpiped and closed in case of error.

Also, the error is correctly propagated, in your case, I believe it will stay tuck if `input` errors.